bokehjs/src/lib/models/widgets/date_picker.ts

```diff
@@ -1,96 +1,75 @@
-import Pikaday from "pikaday"
+import flatpickr from "flatpickr"
 
 import {InputWidget, InputWidgetView} from "./input_widget"
 import {input} from "core/dom"
+import {CalendarPosition} from "core/enums"
 import * as p from "core/properties"
+import {isString, isArray} from "core/util/types"
 import {bk_input} from "styles/widgets/inputs"
-import "styles/widgets/pikaday"
+import "styles/widgets/flatpickr"
 
-Pikaday.prototype.adjustPosition = function(this: Pikaday & {_o: Pikaday.PikadayOptions}): void {
-  if (this._o.container)
-    return
+type DatesList = (string | string[])[]
 
-  this.el.style.position = 'absolute'
-
-  const field = this._o.trigger!
-  const width = this.el.offsetWidth
-  const height = this.el.offsetHeight
-  const viewportWidth = window.innerWidth || document.documentElement!.clientWidth
-  const viewportHeight = window.innerHeight || document.documentElement!.clientHeight
-  const scrollTop = window.pageYOffset || document.body.scrollTop || document.documentElement!.scrollTop
-
-  const clientRect = field.getBoundingClientRect()
-  let left = clientRect.left + window.pageXOffset
-  let top = clientRect.bottom + window.pageYOffset
-
-  // adjust left/top origin to .bk-root
-  left -= this.el.parentElement!.offsetLeft
-  top -= this.el.parentElement!.offsetTop
-
-  // default position is bottom & left
-  if ((this._o.reposition && left + width > viewportWidth) ||
-      (this._o.position!.indexOf('right') > -1 && left - width + field.offsetWidth > 0))
-    left = left - width + field.offsetWidth
-
-  if ((this._o.reposition && top + height > viewportHeight + scrollTop) ||
-      (this._o.position!.indexOf('top') > -1 && top - height - field.offsetHeight > 0))
-    top = top - height - field.offsetHeight
-
-  this.el.style.left = left + 'px'
-  this.el.style.top = top + 'px'
+function _convert_date_list(value: DatesList): any[] {
+  const result: any[] = []
+  for (const item of value) {
+    if (isString(item))
+      result.push(item)
+    else if (isArray(item) && item.length == 2)
+      result.push({from: item[0], to: item[1]})
+  }
+  return result
```

Comment:
> `type DatesList = (string | string[])[]`

Based on the above code the type should be `(string | [string, string])[]`. The second condition is unnecessary. The return type should be `(string | DateLimit<string>)[]`.

Reply:
It would be very meaningful to `type DateType = string`, primarily for readability, even if that doesn't change anything from static typing perspective (TypeScript doesn't have nominal types unfortunately).

Reply:
> `type DateType = string`

Actually, I will introduce that later, because there's an opportunity here to introduce pseudo nominal types.